File: src/main/java/lotto/domain/Result.java
package lotto.domain;


import java.util.Arrays;

public enum Result {

    MISS(0, false, 0),
    FIFTH(3, false, 5000),
    FOURTH(4, false, 50000),
    THIRD(5, false, 1500000),
    SECOND(5, true, 30000000),
    FIRST(6, false, 2000000000);

    private final int matchCount;
    private final boolean bonusMatch;
    private final int price;

    Result(int matchCount, boolean bonusMatch, int price) {
        this.matchCount = matchCount;
        this.bonusMatch = bonusMatch;
        this.price = price;
    }

    public static Result from(int matchCount, boolean bonusMatch) {
        return Arrays.stream(Result.values())
                .filter(r -> r.matchCount == matchCount && r.bonusMatch == bonusMatch)
                .findFirst()
                .orElse(MISS);
    }

    public int getMatchCount() {
        return matchCount;
    }

    public boolean isBonusMatch() {
        return bonusMatch;
    }

    public int getPrice() {
        return price;
    }
}

File: src/test/kotlin/lotto/domain/Result.kt
package lotto.domain

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import kotlin.test.assertEquals

class ResultTest {

    @DisplayName("일치 개수와 보너스 여부에 따라 올바른 등수를 반환한다")
    @ParameterizedTest
    @CsvSource(
        "6, false, FIRST",
        "5, true, SECOND",
        "5, false, THIRD",
        "4, false, FOURTH",
        "3, false, FIFTH",
        "2, true, MISS",
        "0, false, MISS"
    )
    fun `일치 개수와 보너스 여부에 따라 올바른 등수를 반환한다`(
        matchCount: Int,
        bonus: Boolean,
        expected: Result
    ) {
        val result = Result.from(matchCount, bonus)
        assertEquals(expected, result)
    }
}
